Binder Check for Unbound Generics in Methods

kevinsun-dev opened the pull request:
_Based off @RikkiGibson's #43130, a patch for issue #41779._
Thanks to @jpd30 for reporting the issue!

This patch checks for unbound types in the Binder, throwing a CS7003 error without needing to run EmitDiagnostics.

Changelog:

- Added check to Binder_Expression.cs

- Added test for issue #41799.

- Edited tests to throw new error when applicable.

_Sidenote: Apologies for the messy git changes to some of those tests. The newer outputs from the testing suite contained more information in comments that I thought would be worthwhile to leave in._

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:
```diff
@@ -262,9 +262,9 @@ class Test<T>
                 // (17,66): error CS1031: Type expected
                 //         s = nameof(System.Collections.Generic.Dictionary<Program,>.KeyCollection);
                 Diagnostic(ErrorCode.ERR_TypeExpected, ">").WithLocation(17, 66),
-                // (11,27): error CS0305: Using the generic type 'Action<T>' requires 1 type arguments
```
This seems like a worse experience.

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:
```diff
@@ -22918,8 +22953,8 @@ void Test<T>(T p)
                 Diagnostic(ErrorCode.ERR_InvalidExprTerm, "+=").WithArguments("+="),
                 // (9,14): error CS0307: The event 'Program.E' cannot be used with type arguments
                 Diagnostic(ErrorCode.ERR_TypeArgsNotAllowed, "E<>").WithArguments("Program.E", "event"),
-                // (10,19): error CS0307: The event 'Program.E' cannot be used with type arguments
```
This one too.

333fred commented on src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs:
```diff
@@ -6862,5 +6862,38 @@ static void Main(string[] args)
+    }
+}";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (12,18): error CS7003: Unexpected use of an unbound generic name
```
I wouldn't expect CS7003 here, I'd expect something like `Using the generic method 'GetService<T>' requires 1 type arguments`.

AlekseyTs replied:
>// (12,18): error CS7003: Unexpected use of an unbound generic name [](start = 16, length = 67)

I'll create a new thread. I agree with @333fred 

> I wouldn't expect CS7003 here, I'd expect something like "Using the generic method 'GetService<T>' requires 1 type arguments."

333fred commented:
Done review pass (commit 3). This doesn't seem like the right fix to me: It markedly degrades error experiences with unbound generics in some cases and doesn't hugely improve the scenario in the bug. I think we need to use a better error here.

kevinsun-dev commented:
> Done review pass (commit 3). This doesn't seem like the right fix to me: It markedly degrades error experiences with unbound generics in some cases and doesn't hugely improve the scenario in the bug. I think we need to use a better error here.

Edited the check to throw allow binding errors to occur, but adds a diagnostic to catch the new case. All instances where information density was reduced have been corrected, with the only issue being that there's occasionally one more descriptive error in the mix.

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -6062,6 +6062,11 @@ private bool IsUsingAliasInScope(string name)
+                            if (right is GenericNameSyntax genericNameRight && rightHasTypeArguments && typeArgumentsSyntax.Any(SyntaxKind.OmittedTypeArgument) && !IsUnboundTypeAllowed(genericNameRight))
+                            {
+                                diagnostics.Add(ErrorCode.ERR_UnexpectedUnboundGenericName, genericNameRight.Location);
+                            }
```
Nit: newline

kevinsun-dev replied:
Broke the line at a reasonable point to improve readability.

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -6062,6 +6062,11 @@ private bool IsUsingAliasInScope(string name)
                     default:
                         {
+                            // Ensure that there is an error thrown when an Unbound Generic is found.
+                            if (right is GenericNameSyntax genericNameRight && rightHasTypeArguments && typeArgumentsSyntax.Any(SyntaxKind.OmittedTypeArgument) && !IsUnboundTypeAllowed(genericNameRight))
```
Do we need to do this for _every_ `default` case, or just the cases that return from here?

kevinsun-dev replied:
You're right, only the ones that return from here need to be affected. Moved to a more specific scope.

333fred commented:
Done review pass (commit 4)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -6074,6 +6074,13 @@ private bool IsUsingAliasInScope(string name)
                             {
+                                // Ensure that an error is thrown when an unbound generic is found.
+                                if (right is GenericNameSyntax genericNameRight && rightHasTypeArguments &&
+                                    typeArgumentsSyntax.Any(SyntaxKind.OmittedTypeArgument) && !IsUnboundTypeAllowed(genericNameRight))
```
>IsUnboundTypeAllowed [](start = 96, length = 20)

I am confused. It looks like we are binding a generic method invocation, why ```IsUnboundTypeAllowed``` would be relevant here? Also, it is not clear why we check for the error only here, but not, for example, in ```case BoundKind.TypeExpression``` as well. Would it be better to check ```typeArguments``` when we try to bind the invocation?

AlekseyTs replied:
Or perhaps we can check for empty type arguments when we create a BoundMethodGroup node.

---
In reply to: [441136088](https://github.com/dotnet/roslyn/pull/45033#discussion_r441136088) [](ancestors = 441136088)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -6074,6 +6074,13 @@ private bool IsUsingAliasInScope(string name)
+                                if (right is GenericNameSyntax genericNameRight && rightHasTypeArguments &&
+                                    typeArgumentsSyntax.Any(SyntaxKind.OmittedTypeArgument) && !IsUnboundTypeAllowed(genericNameRight))
+                                {
+                                    diagnostics.Add(ErrorCode.ERR_UnexpectedUnboundGenericName, genericNameRight.Location);
```
>genericNameRight [](start = 96, length = 16)

It is not clear why do we think this is a type.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs:
```diff
@@ -3516,81 +3516,108 @@ public void Test()
             var compilation = CreateCompilationWithMscorlib40AndSystemCore(source);
 
             compilation.VerifyDiagnostics(
```
>VerifyDiagnostics [](start = 24, length = 17)

Are we getting more noise in this scenario now?

AlekseyTs commented:
Done with review pass (iteration 5)

kevinsun-dev commented:
Moved check code to `BindInstanceMemberAccess`, the last part of the stack that has error checks when binding method groups. Integrated it into existing `BadArity` check code to report more accurate results. Added new test cases for variants of the original error case.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs:
```diff
@@ -6862,5 +6862,190 @@ static void Main(string[] args)
+
+static class Program
+{
+    static void GetService(this I obj){}
```
>static void GetService(this I obj){} [](start = 4, length = 36)

Is presence of this API necessary? Do we even bind to it in this scenario?

kevinsun-dev replied:
Uncertain, possibly not. Errors don't seem to change with or without the presence of this line.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs:
```diff
@@ -6862,5 +6862,190 @@ static void Main(string[] args)
+    
+    static void M(I provider)
+    {
+        provider.GetService<>();
```
>provider.GetService<>(); [](start = 8, length = 24)

This is not an invocation of a static method. A static method is qualified with a type name rather than with an instance.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs:
```diff
@@ -6862,5 +6862,190 @@ static void Main(string[] args)
+            var source =
+@"interface I
+{
+    object GetService();
```
>object GetService(); [](start = 4, length = 20)

Is presence of this API necessary for the scenario?

kevinsun-dev replied:
The API allows the program to call the GetService generic method, which leads to the error, from what I can tell.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs:
```diff
@@ -6862,5 +6862,190 @@ static void Main(string[] args)
+@"using System;
+class C
+{
+    String GetService(){ return String.Empty; }
```
>String GetService(){ return String.Empty; } [](start = 4, length = 43)

Is presence of this API necessary? Do we even bind to it in this scenario?

AlekseyTs replied:
> I'm fairly certain we do, since removing this line acts as the difference between correct errors and does not emit in SharpLab on master at the moment.

Does the presence of the extension method below make a difference?

---
In reply to: [442352158](https://github.com/dotnet/roslyn/pull/45033#discussion_r442352158) [](ancestors = 442352158)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs:
```diff
@@ -6862,5 +6862,190 @@ static void Main(string[] args)
+@"using System;
+abstract class C
+{
+    String GetService(){ return String.Empty; }
```
>String GetService(){ return String.Empty; } [](start = 4, length = 43)

Is presence of this API necessary? Do we even bind to it in this scenario?

kevinsun-dev replied:
Same as above, this line makes the difference between erroring out and failing emit. If the question is regarding if 'using System' is necessary, the answer is probably no. I was just testing if having actual return types changed results and System.String provided the most convenient object to grab quickly.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs:
```diff
@@ -6862,5 +6862,190 @@ static void Main(string[] args)
+                //         provider.GetService<>();
+                Diagnostic(ErrorCode.ERR_BadArity, "provider.GetService<>").WithArguments("GetService", "method group", "1").WithLocation(15, 9)
+            );
+        }
```
> } [](start = 7, length = 2)

I recommend splitting test scenarios in three category:
1) binding to an instance method
2) binding to an extension method
3) binding to a static method

Test all of them separately for generic and non-generic cases.

AlekseyTs replied:
If a combination of both an instance and an extension method is needed in order to observe the problem, then we need to keep tests like that. However, I would like us to get a better understanding what kind of interaction is going on that causes us to fail in a scenario like that. The error messages indicate that we bound to the extension method. Why would presence of some other method cause us to fail to report an error.

---
In reply to: [442351010](https://github.com/dotnet/roslyn/pull/45033#discussion_r442351010) [](ancestors = 442351010)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/UnboundGenericType.vb:
```diff
@@ -289,6 +289,51 @@ End Class
+                              ~
+                </expected>)
+
+        End Sub
```
>End Sub [](start = 8, length = 7)

I recommend using the same test matrix as we end up with for C#.

kevinsun-dev replied:
As in using all the same tests (translated to VB), or using the same testing methodology?

AlekseyTs replied:
> As in using all the same tests (translated to VB), or using the same testing methodology?

I would port all the same tests. But let's wait until we are good on C# side.

---
In reply to: [442359826](https://github.com/dotnet/roslyn/pull/45033#discussion_r442359826) [](ancestors = 442359826)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs:
```diff
@@ -6862,5 +6862,190 @@ static void Main(string[] args)
+        public void Bug41779_Static()
+        {
+            var source =
+@"interface I
```
>interface I [](start = 2, length = 11)

Does this have to be an interface? If not, let's use class to eliminate the DIM requirement

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -5942,9 +5942,9 @@ private bool IsUsingAliasInScope(string name)
                 }
 
-                var typeArgumentsSyntax = right.Kind() == SyntaxKind.GenericName ? ((GenericNameSyntax)right).TypeArgumentList.Arguments : default(SeparatedSyntaxList<TypeSyntax>);
+                SeparatedSyntaxList<TypeSyntax> typeArgumentsSyntax = right is GenericNameSyntax genericName ? genericName.TypeArgumentList.Arguments : default;
```
>SeparatedSyntaxList<TypeSyntax> typeArgumentsSyntax = right is GenericNameSyntax genericName ? genericName.TypeArgumentList.Arguments : default; [](start = 16, length = 144)

Is there a change in behavior in this function? If not, please revert the refactoring.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -6207,7 +6207,9 @@ private BoundExpression MakeMemberAccessValue(BoundExpression expr, DiagnosticBa
 
-                    if (!boundMethodGroup.HasErrors && boundMethodGroup.ResultKind == LookupResultKind.Empty && typeArgumentsSyntax.Any(SyntaxKind.OmittedTypeArgument))
+                    if (!boundMethodGroup.HasErrors && (boundMethodGroup.ResultKind == LookupResultKind.Empty ||
+                        (right is GenericNameSyntax genericNameRight && !IsUnboundTypeAllowed(genericNameRight))) &&
```
>!IsUnboundTypeAllowed(genericNameRight) [](start = 72, length = 39)

Why do we need this condition, we are not binding a type?

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -1221,7 +1221,7 @@
     <value>Cannot apply attribute class '{0}' because it is generic</value>
   </data>
-  <data name="ERR_DuplicateBound" xml:space="preserve">
+  <data name="ERR_DuplicateBound" xml:space="preserve">                                
```
>  <data name="ERR_DuplicateBound" xml:space="preserve">          [](start = 0, length = 64)

It looks like there are no meaningful changes in this file. Consider reverting

kevinsun-dev replied:
Will do

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -6207,7 +6207,9 @@ private BoundExpression MakeMemberAccessValue(BoundExpression expr, DiagnosticBa
+                        (right is GenericNameSyntax genericNameRight && !IsUnboundTypeAllowed(genericNameRight))) &&
+                        typeArgumentsSyntax.Any(SyntaxKind.OmittedTypeArgument))
                     {
                         Error(diagnostics, ErrorCode.ERR_BadArity, node, rightName, MessageID.IDS_MethodGroup.Localize(), typeArgumentsSyntax.Count);
```
>typeArgumentsSyntax.Count [](start = 122, length = 25)

I see. It looks like the error message isn't really based on any actual method. It is based solely on syntax.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -6207,7 +6207,9 @@ private BoundExpression MakeMemberAccessValue(BoundExpression expr, DiagnosticBa
                         flags);
 
-                    if (!boundMethodGroup.HasErrors && boundMethodGroup.ResultKind == LookupResultKind.Empty && typeArgumentsSyntax.Any(SyntaxKind.OmittedTypeArgument))
+                    if (!boundMethodGroup.HasErrors && (boundMethodGroup.ResultKind == LookupResultKind.Empty ||
```
>boundMethodGroup.ResultKind == LookupResultKind.Empty [](start = 56, length = 53)

Rather than adding something to this condition, could we simply remove it? BTW, what is the ```boundMethodGroup.ResultKind``` for our scenario?

kevinsun-dev replied:
The `boundMethodGroup.ResultKind` for this scenario is `WrongArity`, huh. It seems that some check may actually be in place. Going to modify this to check for that condition instead to see if it works.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs:
```diff
@@ -3528,55 +3528,55 @@ public void Test()
                 // (18,13): error CS0815: Cannot assign method group to an implicitly-typed variable
                 //         var omittedArgFunc1 = "string literal".ExtensionMethod1<>;
                 Diagnostic(ErrorCode.ERR_ImplicitlyTypedVariableAssignedBadValue, @"omittedArgFunc1 = ""string literal"".ExtensionMethod1<>").WithArguments("method group").WithLocation(18, 13),
+                // (18,31): error CS0305: Using the generic method group 'ExtensionMethod1' requires 1 type arguments
```
>// (18,31): error CS0305: Using the generic method group 'ExtensionMethod1' requires 1 type arguments [](start = 16, length = 101)

It looks like there are no meaningful changes in this file. Consider reverting. Otherwise, please move this error to its original place so that reviewers could easily see that there are no changes in behavior.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:
```diff
@@ -22798,77 +22798,124 @@ void Test(object p)
+                    // (21,14): error CS0308: The non-generic method 'Program.M()' cannot be used with type arguments
+                    //         Test(M<>());
+                    Diagnostic(ErrorCode.ERR_HasNoTypeVars, "M<>").WithArguments("Program.M()", "method").WithLocation(21, 14),
+                    // (23,19): error CS7003: Unexpected use of an unbound generic name
```
>// (23,19): error CS7003: Unexpected use of an unbound generic name [](start = 20, length = 67)

This feels like a noise

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:
```diff
@@ -22798,77 +22798,124 @@ void Test(object p)
+                    // (23,19): error CS0307: The field 'Program.f' cannot be used with type arguments
+                    //         Test(this.f<>);
+                    Diagnostic(ErrorCode.ERR_TypeArgsNotAllowed, "f<>").WithArguments("Program.f", "field").WithLocation(23, 19),
+                    // (24,19): error CS7003: Unexpected use of an unbound generic name
```
>// (24,19): error CS7003: Unexpected use of an unbound generic name [](start = 20, length = 67)

This feels like a noise

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:
```diff
@@ -22798,77 +22798,124 @@ void Test(object p)
+                    // (24,19): error CS0307: The property 'Program.P' cannot be used with type arguments
+                    //         Test(this.P<>);
+                    Diagnostic(ErrorCode.ERR_TypeArgsNotAllowed, "P<>").WithArguments("Program.P", "property").WithLocation(24, 19),
+                    // (25,19): error CS7003: Unexpected use of an unbound generic name
```
>// (25,19): error CS7003: Unexpected use of an unbound generic name [](start = 20, length = 67)

This feels like a noise

kevinsun-dev commented on src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:
```diff
@@ -22798,77 +22798,124 @@ void Test(object p)
+                    // (29,13): error CS0308: The non-generic method 'Program.M()' cannot be used with type arguments
+                    //         m = M<>;
+                    Diagnostic(ErrorCode.ERR_HasNoTypeVars, "M<>").WithArguments("Program.M()", "method").WithLocation(29, 13),
+                    // (30,18): error CS7003: Unexpected use of an unbound generic name
```
these 7003 errors shouldn't occur anymore, now that the code has been removed. I must have missed this test as something that shouldn't change. Will revert.

AlekseyTs commented:
Done with review pass (iteration 6). Implementation looks good to me. Consider adding recommended tests.

kevinsun-dev commented:
Still getting used to CodeFlow, couldn't seem to find a way to drop a comment as the PR Author, gonna ask about that.
Expanded testing to include binding to generics and non-generics for instance, extension, and static methods, as well as testing methods requiring no generic types, one generic type, and two generic types for all cases. Added translated test matrix to VB as well.

333fred commented on src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/UnboundGenericType.vb:
```diff
@@ -289,6 +289,331 @@ End Class
+                      ~
+            </expected>)
+
+        End Sub
```
Nit: extra newline below.
